format/FormatNexusJungfrauExt.py

```diff
@@ -0,0 +1,123 @@
+import h5py
+import numpy as np
+
+from xfel.util.jungfrau import pad_stacked_format
+
+from dials.array_family import flex
+
+from dxtbx.format.FormatNexus import FormatNexus
+
+"""
+This class can read a Jungfrau image file that includes, in addition to the corrected data,
+the gain maps, and the dark (pedestal) averages and RMS, providing the user access
+to that data.
+This was written to process SWISSFEL image files created using the CCTBX script jf16m_cxigeom2nexus.py
+
+  libtbx.python cctbx_project/xfel/swissfel/jf16m_cxigeom2nexus.py \
+  unassembled_file=X \
+  raw_file=Y \
+  geom_file=Z \
+  pedestal_file=P \
+  gain_file=G \
+  detector_distance=D \
+  include_spectra=True \
+  output_file=O \
+  raw=False \
+  beam_file=B
+
+where X,Y are the corrected,raw hdf5 files obtained at SWISSFEL for a particular run, Z is the crystFEL geom file
+provided by SWISSFEL for the jungfrau, P is the hdf5 file provided by swissfel containing the pedestal data,  G is
+the gainmap hdf5 file provided by SWISSFEL, D is the detector distance in mm, O is an output filename (to be read by
+this format class) and B is the hdf5 file provided by SWISSFEL containing the wavelength spectra measurements (optional)
+"""
+
+
+class FormatNexusJungfrauExt(FormatNexus):
+    @staticmethod
+    def understand(image_file):
+        with h5py.File(image_file, "r") as handle:
+            try:
+                data = handle["/entry/data"]
+                if "pedestal" in data and "gains" in data and "raw" in data:
+                    return True
+            except (KeyError, AttributeError):
+                pass
+        return False
+
+    def _slice_correction_array(self, array):
+        """
+        :param array: numpy array where the slow axis is image number
+        and e.g. array[i] corresponds to the ith nexus data entry
+        :return: array sliced according to the nexus geometry specification
+        """
+        return np.array(
+            [
+                array[:, self._slices[i][0], self._slices[i][1]]
+                for i in range(len(self._slices))
+            ]
+        )
+
+    def _start(self):
+        super(FormatNexusJungfrauExt, self)._start()
+        data_handle = self._reader.entries[0].data[0].handle
+
+        # nexus slices
+        self._slices = self._raw_data._datalists[0]._all_slices
+
+        self._gain = self._slice_correction_array(data_handle["gains"][()])
+        self._pedestal = self._slice_correction_array(data_handle["pedestal"][()])
+        self._pedestalRMS = self._slice_correction_array(data_handle["pedestalRMS"][()])
+
+        self.dxtbx_detector_shape = self._gain[:, 0].shape
+
+    def get_14bit_component(self, index, as_flex=False):
```

Comment:
I'd like this to evolve more generally into `get_raw_data()` since _this is the raw data_

(just a comment)

Reply:
It is, but without a way to apply corrections in this format class, it's not very useful.  The corrections are (data-pedestal)/gain/keV to get photon counts.

We have an alternate version of this format class that we made first, for a version of these master files that only had the raw data in it.  That version applied the corrections in memory.  We found it was too slow to be practical so [we took it out](https://github.com/cctbx/dxtbx/pull/303/commits/92bb23d11f2bca53c80ad0838b1e0d7e90902d3f) in favor of this version.  This version has a link to the corrected data for faster performance, and to the raw data so we can pull out gain maps and pedestal RMS and what have you.

We did look into the dxtbx machinery to provide a dynamic gain and pedestal in imageset's [get_corrected_data](https://github.com/cctbx/dxtbx/blob/master/imageset.h#L646), but there isn't a clean way to do it on a per-image basis.  Regardless, it still would have been slow.

Reply:
Yea, I also thought about utilizing the built-in imageset methods, as that would be most ideal. I think the trouble with that approach for this format is the 3 pedestals and 3 gain modes that fluctuate from shot to shot. it seems that get_pedestal and get_gain (as part of the get_corrected_data in dxtbx/imageset.h) wont work without major intervention  ?

Reply:
Right, because they are external lookups that don't call into the format class.